[KEP-4639] Add `ImageVolumeSource` API

saschagrunert opened the pull request:
#### What type of PR is this?

/kind feature


#### What this PR does / why we need it:

Adding the required Kubernetes API so that the kubelet can start using it. This patch also adds the corresponding alpha feature gate as outlined in KEP 4639.

#### Which issue(s) this PR fixes:

Refers to https://github.com/kubernetes/enhancements/issues/4639

#### Special notes for your reviewer:

None

#### Does this PR introduce a user-facing change?

```release-note
Revised the Pod API with alpha support for volumes derived from OCI artifacts. This feature is behind the `ImageVolume` feature gate.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
KEP: https://github.com/kubernetes/enhancements/issues/4639
```

saschagrunert committed:
Add ImageVolumeSource API

Adding the required Kubernetes API so that the kubelet can start using
it. This patch also adds the corresponding alpha feature gate as
outlined in KEP 4639.

Signed-off-by: Sascha Grunert <sgrunert@redhat.com>

ffromani commented:
/triage accepted
/priority important-soon

sftim commented:
Changelog suggestion
```diff
-Added `OCIVolumeSource` API and `OCIVolume` alpha feature gate.
+Revised the Pod API with alpha support for volumes derived from OCI artefacts.
+This feature is behind the `OCIVolume` feature gate.
```

You won't be able to `kubectl get OCIVolumeSources` after this merges; the changelog as I see it now could be read as implying that we've added a new API kind.

saschagrunert commented:
> Changelog suggestion
> 
> ```diff
> -Added `OCIVolumeSource` API and `OCIVolume` alpha feature gate.
> +Revised the Pod API with alpha support for volumes derived from OCI artefacts.
> +This feature is behind the `OCIVolume` feature gate.
> ```
> 
> You won't be able to `kubectl get OCIVolumeSources` after this merges; the changelog as I see it now could be read as implying that we've added a new API kind.

Thank you, updated the release note as suggested.

saschagrunert commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -7624,3 +7628,14 @@ const (
+// OCIVolumeSource represents a OCI volume resource.
+type OCIVolumeSource struct {
+	// Required: Image or artifact reference to be used.
+	Reference string `json:"reference,omitempty" protobuf:"bytes,1,opt,name=reference"`
```
Thanks, fixed as suggested :+1:

saschagrunert commented:
/retest

saschagrunert commented:
/retest

saschagrunert commented:
/test pull-kubernetes-integration

@kubernetes/sig-api-machinery-api-reviews PTAL

saschagrunert commented:
/retest

mrunalp commented:
@smarterclayton ptal

saschagrunert commented:
/assign @smarterclayton

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -181,6 +181,10 @@ type VolumeSource struct {
+	// oci represents a OCI object (image or artifact) pulled and mounted on the kubelet's host machine.
+	// +featureGate=OCIVolume
+	// +optional
+	OCI *OCIVolumeSource `json:"oci,omitempty" protobuf:"bytes,30,opt,name=oci"`
```
I am torn on whether to argue that this should be called `Image` - historically that would be correct, since Kubernetes does not require the CRI to implement OCI images (you could be based on docker v2 spec, or implement a transformation layer).

saschagrunert replied:
My personal opinion is that `oci` fits greatly for both, images and artifacts. Maybe also `ociObject` if it's not specific enough.

cc @sallyom @mrunalp @haircommander

mrunalp replied:
I had similar feelings when I saw this name first compared to existing `Image`. I think naming it `Image` will make it more future proof if we want to add support for some other tranport later on.

saschagrunert replied:
Makes sense, changed the image name to `Image`.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -7645,3 +7649,14 @@ const (
+	// Required: Image or artifact reference to be used.
+	Reference string `json:"reference" protobuf:"bytes,1,opt,name=reference"`
+
+	// Policy for pulling OCI objects.
```
Describe how each of the policies works in practice in the godoc, so a user knows what Always means

saschagrunert replied:
Added all possible values and docs as suggested.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -181,6 +181,10 @@ type VolumeSource struct {
 	//
 	// +optional
 	Ephemeral *EphemeralVolumeSource `json:"ephemeral,omitempty" protobuf:"bytes,29,opt,name=ephemeral"`
+	// oci represents a OCI object (image or artifact) pulled and mounted on the kubelet's host machine.
```
Godoc should include:

* Describe when the volume is resolved (at pod startup) and describe how the policy impacts that resolution.
* Describe whether the volume is reresolved.
* Describe that the OCI object will be mounted as a directory.
* Describe that the volume will be read only.

saschagrunert replied:
Added all the docs as requested.

smarterclayton commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2920,6 +2928,18 @@ func ValidateVolumeMounts(mounts []core.VolumeMount, voldevices map[string]strin
 		}
 
+		// Disallow subPath/subPathExpr for OCI volumes
+		if utilfeature.DefaultFeatureGate.Enabled(features.OCIVolume) {
```
Be sure to include this in the api doc / godoc of the volume source - `subPath mounts of this volume source are not supported.`.

saschagrunert replied:
Done.

saschagrunert commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -7645,3 +7649,14 @@ const (
+// OCIVolumeSource represents a OCI volume resource.
+type OCIVolumeSource struct {
+	// Required: Image or artifact reference to be used.
+	Reference string `json:"reference" protobuf:"bytes,1,opt,name=reference"`
```
Done

saschagrunert commented:
/test pull-kubernetes-integration

saschagrunert commented:
/test pull-kubernetes-integration

saschagrunert commented:
/test pull-kubernetes-e2e-gce

mrunalp commented on pkg/apis/core/types.go:
```diff
@@ -6493,3 +6508,19 @@ const (
 )
+
+// OCIVolumeSource represents a OCI volume resource.
+type OCIVolumeSource struct {
```
I think this should be ImageVolumeSource as well.

cc: @smarterclayton for thoughts.

saschagrunert replied:
Renamed that as well together with the feature gate. It's a separate commit on top of this PR which can also be dropped again.

smarterclayton replied:
I think `image` is the right name, given that we don't have a strong spec to reference for artifacts, while we do for behavior of images in Kubernetes.  

I added some suggested godoc edits to leave room for the container runtime implementation to add additional sources (as long as they are consistent with the behavior of container image mounting).  Single file mounting (vs directory) is the potentially only limitation I can see, but we could always define a CRI level transformation for any future OCI artifact spec representing single files or variation by adding new fields to the volume source to control a mapping from source -> directory.

I think we'd need a really strong reason to be inconsistent with container image, given we're expecting to reuse the CRI implementation of pull (same behavior), the fact that the vast majority of content here is going to be in the form of OCI images, and that users will have less to understand if image source pairs well with container image.

So +1 from me.

saschagrunert replied:
Thanks, marking this point as resolved since we reached agreement.

saschagrunert commented:
/retest

haircommander commented on pkg/api/pod/util.go:
```diff
@@ -1260,3 +1261,56 @@ func MarkPodProposedForResize(oldPod, newPod *api.Pod) {
+
+// KEP: https://kep.k8s.io/4639
+func dropImageVolumes(podSpec, oldPodSpec *api.PodSpec) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume) || imageVolumesInUse(oldPodSpec) {
```
shouldn't this be !imageVolumesInUse ?

saschagrunert replied:
Hm, no the logic seems to be correct looking at the other functions and also test cases. Am I missing something?

haircommander replied:
if featureGate: don't drop
if !imageVolumesInUse: nothing to drop
if either are not set, then don' t we need to drop? as in: if image volumes are being used but the feature is disabled, wouldn't we need to clear them?

saschagrunert replied:
Hm, I was looking how others implemented it and it’s the same logic for `dropPodLifecycleSleepAction`, right?

https://github.com/kubernetes/kubernetes/blob/14592ef39d6c08fb5af48df3edadb4556dee2e9a/pkg/api/pod/util.go#L720

haircommander replied:
that one also looks wrong haha for instance here's userns https://github.com/kubernetes/kubernetes/blob/14592ef39d6c08fb5af48df3edadb4556dee2e9a/pkg/api/pod/util.go#L650

haircommander replied:
the majority of them check !inUse AFAICT

saschagrunert replied:
We drop if `!enabled && !inuse`, means we don’t drop if `!(!enabled && !inuse)` which is the same as `enabled || inuse`, right? Am I wrong?

haircommander replied:
okay you're right, sorry for the noise. I think I still don't understand why this is the way this code works, but you're definitely following precedent and it must be there for a reason I'm missing :upside_down_face:

saschagrunert replied:
It’s tricky but I was aligning with the existing logic and test cases. So I assume that’s correct. 👍

haircommander commented on pkg/apis/core/types.go:
```diff
@@ -6493,3 +6507,19 @@ const (
+type ImageVolumeSource struct {
+	// Required: Image or artifact reference to be used.
+	// Behaves in the same way as pod.spec.containers[*].image.
+	// More info: https://kubernetes.io/docs/concepts/containers/images
```
should we add `// +required` here?

saschagrunert replied:
It does not change the public API if we add that comment, neither to this file nor `staging/src/k8s.io/api/core/v1/types.go`. So I'd prefer to keep the `Required:` in the comment.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -181,6 +181,20 @@ type VolumeSource struct {
+	// - Never: the kubelet never pulls the reference and only uses a local image or artifact. Container creation will fail if the reference isn't present.
+	// - IfNotPresent: the kubelet pulls if the reference isn't already present on disk. Container creation will fail if the reference isn't present and the pull fails.
+	//
+	// The volume gets re-resolved if the pod gets recreated, which means that new remote content will become available on pod recreation.
```
Trying to think about whether to add any additional description, but this is pretty succinct.  

We don't really define the retry behavior of volume start well, but given that this one is likely to add both latency and more failure causes, we should probably add some content along those lines as an additional sentence:

> A failure to resolve or pull the image during pod startup will block containers from starting and may add significant latency.  Failures will be retried using normal volume backoff and will be reported on the pod reason and message. [need to verify that is true for this change]

It would be nice (depending on how complex to implement) if failures show up in both the PodInitialized condition.  It's probably required for those errors to show up in pod.status.reason / message (I think they do today for pull failures but would like to double check during implementation).

saschagrunert replied:
Yes, we're pulling in the same manner during pod sync and will report the same errors. I updated the docs to reflect that.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -181,6 +181,20 @@ type VolumeSource struct {
 	//
 	// +optional
 	Ephemeral *EphemeralVolumeSource `json:"ephemeral,omitempty" protobuf:"bytes,29,opt,name=ephemeral"`
+	// image represents a OCI object (image or artifact) pulled and mounted on the kubelet's host machine.
```
Thinking about the guarantees we provide here, I want to suggest a sentence to include here that speaks to the consistency with the image field on containers as well as to the responsibility of the CRI implementation to resolve the details:

> image represents an OCI object (a container image or artifact) pulled and mounted on the kubelet's host machine.  

and then later on (maybe right before `The OCI object gets mounted ...`):

> The types of objects that may be mounted by this volume are defined by the container runtime implementation on a host machine and at minimum must include all valid types supported by the container image field.

Which I think is an important nuance - we want to leave it open for extension by distros, by OCI, and if someone today has a custom CRI that interprets this value it should be valid for them to do so.

saschagrunert replied:
Alright, updated the docs as suggested. :+1:

saschagrunert replied:
Alright, updated the docs as suggested. :+1:

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -7645,3 +7659,19 @@ const (
+	// Always: the kubelet always attempts to pull the reference. Container creation will fail If the pull fails.
+	// Never: the kubelet never pulls the reference and only uses a local image or artifact. Container creation will fail if the reference isn't present.
+	// IfNotPresent: the kubelet pulls if the reference isn't already present on disk. Container creation will fail if the reference isn't present and the pull fails.
+	// Defaults to IfNotPresent
```
A point of inconsistency with containers image - the pull policy for a container is overriden to "Always" if a latest tag is used (i.e. `// Defaults to Always if :latest tag is specified, or IfNotPresent otherwise.`).

While this behavior is confusing, I think it would be more confusing if we were not consistent with the container image field in this fashion.  We would need a really strong argument to *not* be consistent - while I tried to find one (I don't like the behavior), I could not generate one that was strong enough to create a point of inconsistency.

For that reason, I think we need to have the same behavior and the same godoc here.

mrunalp replied:
yeah, agree.

saschagrunert replied:
Makes sense, I updated the docs wrt that detail.

smarterclayton commented:
Added a second round of comments, the godoc is looking really good and I think this is very close to approval.

saschagrunert commented:
> Added a second round of comments, the godoc is looking really good and I think this is very close to approval.

Thank you! I updated the docs since the suggested additions were extremely useful. Do you mind taking a look again?

saschagrunert commented:
/test pull-kubernetes-e2e-kind

saschagrunert commented:
Rebased a few times to resolve the conflicts.

smarterclayton commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -69,6 +69,9 @@ func SetDefaults_Volume(obj *v1.Volume) {
 			EmptyDir: &v1.EmptyDirVolumeSource{},
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume) && obj.Image != nil && obj.Image.PullPolicy == "" {
```
We need the "latest" defaulting behavior if image pull policy is empty, and test case.

saschagrunert replied:
Thanks, fixed as suggested :+1:

mrunalp commented:
@deads2k for another review and lgtm

deads2k commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -181,6 +181,22 @@ type VolumeSource struct {
+	// - Never: the kubelet never pulls the reference and only uses a local image or artifact. Container creation will fail if the reference isn't present.
+	// - IfNotPresent: the kubelet pulls if the reference isn't already present on disk. Container creation will fail if the reference isn't present and the pull fails.
+	//
+	// The volume gets re-resolved if the pod gets recreated, which means that new remote content will become available on pod recreation.
```
Since pods are mostly immutable, what does it mean to have a pod recreated?

smarterclayton replied:
Good point, I "knew" that meant delete and recreated, but many users won't.

saschagrunert replied:
Slightly modified the comment to include container deletion.

deads2k commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -7662,3 +7678,19 @@ const (
 )
+
+// ImageVolumeSource represents a image volume resource.
+type ImageVolumeSource struct {
```
If I want to use a specific pull secret for this image, how do I specify that?  I'm guessing the "normal" keyring made up of node credentials, SA image pull secrets, and podspec image pull secrets, with each potentially valid credential being used.  But we should be clear.

I don't see a need or value in subdividing further at this moment.

saschagrunert replied:
I added a comment to clarify that, thank you!

deads2k commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -7662,3 +7678,19 @@ const (
+	// Required: Image or artifact reference to be used.
+	// Behaves in the same way as pod.spec.containers[*].image.
+	// More info: https://kubernetes.io/docs/concepts/containers/images
+	Reference string `json:"reference" protobuf:"bytes,1,opt,name=reference"`
```
a container's image field is optional to allow injection from higher level construct/controllers when embedded inside of say a deployment.  Do the same forces apply?  if so, update this to optional.  if not, please explain why.

saschagrunert replied:
Should become optional then, updated the field.

smarterclayton replied:
I think they could, since the content could be provided by a higher level injection.  Good catch, I missed that.

deads2k commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -7662,3 +7678,21 @@ const (
+	// Required: Image or artifact reference to be used.
+	// Behaves in the same way as pod.spec.containers[*].image.
+	// Pull secrets will be assembled in the same way as for the container image by looking up node credentials, SA image pull secrets and pod spec image pull secrets.
+	// More info: https://kubernetes.io/docs/concepts/containers/images
```
```suggestion
	// More info: https://kubernetes.io/docs/concepts/containers/images
	// This field is optional to allow higher level config management to default or override
	// container images in workload controllers like Deployments and StatefulSets.
```

saschagrunert replied:
Thanks, applied as suggested.

deads2k commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -7662,3 +7678,21 @@ const (
+type ImageVolumeSource struct {
+	// Required: Image or artifact reference to be used.
+	// Behaves in the same way as pod.spec.containers[*].image.
+	// Pull secrets will be assembled in the same way as for the container image by looking up node credentials, SA image pull secrets and pod spec image pull secrets.
```
```suggestion
	// Pull secrets will be assembled in the same way as for the container image by looking up node credentials, SA image pull secrets, and pod spec image pull secrets.
```

the kube project believes in the oxford comma

saschagrunert replied:
Thanks, applied as suggested.

deads2k commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -181,6 +181,22 @@ type VolumeSource struct {
+	// The volume is resolved at pod startup depending on which PullPolicy value is provided:
+	//
+	// - Always: the kubelet always attempts to pull the reference. Container creation will fail If the pull fails.
+	// - Never: the kubelet never pulls the reference and only uses a local image or artifact. Container creation will fail if the reference isn't present.
```
is there any case when the container image would be re-pulled by the kubelet and the volume image would not be re-pulled?

is there any case where the container image would not be re-pulled by the kubelet and the volume image would be re-pulled?

saschagrunert replied:
> is there any case when the container image would be re-pulled by the kubelet and the volume image would not be re-pulled?

If the container image is set to a policy of `Always` and the volume image to `Never` (or `IfNotPresent` and existing on disk).

> is there any case where the container image would not be re-pulled by the kubelet and the volume image would be re-pulled?

The other way around: If the volume image is set to a policy of `Always` and the container image to `Never` (or `IfNotPresent` and existing on disk).

deads2k commented on pkg/apis/core/validation/validation.go:
```diff
@@ -732,6 +732,14 @@ func validateVolumeSource(source *core.VolumeSource, fldPath *field.Path, volNam
 			}
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume) && source.Image != nil {
```
validation functions aren't supposed to directly depend on featuregates.  Instead, wire an options struct like is done for jobs here: https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/batch/job/strategy.go#L346-L347 and https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/batch/validation/validation.go#L152 .

saschagrunert replied:
Yep, fixed as suggested.

deads2k commented on pkg/apis/core/validation/validation.go:
```diff
@@ -8172,3 +8192,12 @@ func validateLinuxContainerUser(linuxContainerUser *core.LinuxContainerUser, fld
+
+func validateImageVolumeSource(imageVolume *core.ImageVolumeSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(imageVolume.Reference) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("reference"), ""))
+	}
```
this should be required for pods and allowed to be empty for deployments (other things containing podspec).

Please add tests for both cases.

saschagrunert replied:
Fixed using `ResourceIsPod` option.

deads2k commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2920,6 +2928,18 @@ func ValidateVolumeMounts(mounts []core.VolumeMount, voldevices map[string]strin
+		// Disallow subPath/subPathExpr for image volumes
+		if utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume) {
+			if v, ok := volumes[mnt.Name]; ok && v.Image != nil {
+				if mnt.SubPath != "" {
```
kube standard is `len(mnt.SubPath) != 0`.  Please update here and below.

saschagrunert replied:
Fixed as suggested.

deads2k commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -69,6 +70,14 @@ func SetDefaults_Volume(obj *v1.Volume) {
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume) && obj.Image != nil && obj.Image.PullPolicy == "" {
+		// PullPolicy defaults to Always if :latest tag is specified, or IfNotPresent otherwise.
+		if strings.HasSuffix(obj.Image.Reference, ":latest") {
```
use the same `parsers.ParseImageName` as above in this file please.

saschagrunert replied:
Done

deads2k commented on pkg/api/pod/util.go:
```diff
@@ -1260,3 +1261,56 @@ func MarkPodProposedForResize(oldPod, newPod *api.Pod) {
+	}
+	podSpec.Volumes = newVolumes
+
+	dropVolumeMounts := func(givenMounts []api.VolumeMount) (newVolumeMounts []api.VolumeMount) {
+		for _, m := range givenMounts {
+			if !imageVolumeNames.Has(m.Name) {
+				newVolumeMounts = append(newVolumeMounts, m)
+			}
+		}
+		return newVolumeMounts
+	}
```
No change needed here, but thinking to myself that in this case, I'm not certain this is kinder than letting the pod fail validation.

saschagrunert replied:
I'm more for simplicity and decided to remove that code + tests from [c5a8de6 ](https://github.com/kubernetes/kubernetes/commit/c5a8de67a91d0de996ac39dbd625285d60e97537)to [eecdd1d](https://github.com/kubernetes/kubernetes/commit/eecdd1ddfe0f8aeee5b0693eb8ae17c20fad2847)

deads2k replied:
> No change needed here, but thinking to myself that in this case, I'm not certain this is kinder than letting the pod fail validation.

Miscommunication here.  There are two parts

1. Dropping the volume itself.  This must be done.
2. Dropping the mounts that reference that volume.  It helps make the pod valid, it probably doesn't make the pod succeed overall but it passes validation.  I could go either way on this part, so I wasn't hoping for a change.  On detailed reflection, I can imagine writing a cross-compatible podspec with a binary that looks in location/A for data before downloading it from another source, so this should be restored.

saschagrunert replied:
Makes sense, I restored the code.

deads2k commented:
/lgtm
/approve

gnufied commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2920,6 +2928,18 @@ func ValidateVolumeMounts(mounts []core.VolumeMount, voldevices map[string]strin
+		if opts.AllowImageVolumeSource {
+			if v, ok := volumes[mnt.Name]; ok && v.Image != nil {
+				if len(mnt.SubPath) != 0 {
+					allErrs = append(allErrs, field.Invalid(idxPath.Child("subPath"), mnt.SubPath, "not allowed in image volume sources"))
```
IMO - it should be specified in api docs that, using `subpath` for this volume type is not permitted.

saschagrunert replied:
Is this the line you're looking for? https://github.com/kubernetes/kubernetes/blob/0caeba5cbe7793d09586556e056b08f4a8e0adba/pkg/apis/core/types.go#L199

gnufied replied:
I would have thought, it will be also documented in `VolumeMounts` field doc.

saschagrunert replied:
Hm, I don't think that `VolumeMount` is the right place to document that. :thinking:

gnufied commented on pkg/apis/core/types.go:
```diff
@@ -6508,3 +6524,23 @@ const (
+	// Defaults to Always if :latest tag is specified, or IfNotPresent otherwise.
+	// +optional
+	PullPolicy PullPolicy
+}
```
It should be mentioned that - pod's `fsGroupChangePolicy` field will have no effect for this volume type.

saschagrunert replied:
Yep, see: https://github.com/kubernetes/kubernetes/pull/126281

thockin commented:
One thing I missed why is subPath explicitly disallowed?

saschagrunert commented:
> One thing I missed why is subPath explicitly disallowed?

We did that for simplicity. I'll put it on the list for the beta graduation.
